client/client.go

```diff
@@ -383,25 +393,45 @@ func (r *NotaryRepository) Publish() error {
 		return err
 	}
 
+	// these are the tuf files we will need to update, serialized as JSON before
+	// we send anything to remote
+	update := make(map[string][]byte)
+
 	// check if our root file is nearing expiry. Resign if it is.
-	if nearExpiry(r.tufRepo.Root) || r.tufRepo.Root.Dirty {
-		if err != nil {
-			return err
-		}
+	if nearExpiry(r.tufRepo.Root) || r.tufRepo.Root.Dirty || updateRoot {
 		root, err = r.tufRepo.SignRoot(data.DefaultExpires("root"))
+
+		rootJSON, err := json.Marshal(root)
 		if err != nil {
 			return err
 		}
-		updateRoot = true
+		update[data.CanonicalRootRole] = rootJSON
 	}
-	// we will always resign targets and snapshots
-	targets, err := r.tufRepo.SignTargets("targets", data.DefaultExpires("targets"))
+
+	// we will always resign targets, and snapshots if we have the snapshots key
+	targetsJSON, err := serializeCanonicalRole(r.tufRepo, data.CanonicalTargetsRole)
 	if err != nil {
 		return err
 	}
-	snapshot, err := r.tufRepo.SignSnapshot(data.DefaultExpires("snapshot"))
-	if err != nil {
-		return err
+	update[data.CanonicalTargetsRole] = targetsJSON
+
+	// do not update the snapshot role if we do not have the snapshot key or
+	// any snapshot data.  There might not be any snapshot data the repo was
+	// initialized with the snapshot signing role delegated to the server.
+	// The repo might have snapshot data, because it was requested from
+	// the server by listing, but not have the snapshot key, so signing will
+	// fail.
+	if r.tufRepo.Snapshot != nil {
+		snapshotJSON, err := serializeCanonicalRole(
+			r.tufRepo, data.CanonicalSnapshotRole)
+		if err == nil { // we have the key - snapshot signed, let's update it
+			update[data.CanonicalSnapshotRole] = snapshotJSON
+		} else if _, ok := err.(signed.ErrNoKeys); ok {
+			logrus.Debugf("Client does not have the key to sign snapshot. " +
```

Comment:
You bring up a good point though - if we designed the server as the signer, then we don't generate a snapshot file.  If we then change it again so that we have the key, no snapshot file can be generated, so we might not be able to switch back without having first published.  I have tests for switching back and forth in #348 (https://github.com/docker/notary/pull/348/files#diff-fce84029d346504e44323c30d21a3808R933) but not for publishing.  Will add that.

The "+" was to wrap a long text line.